### PR TITLE
Add Garden provider

### PR DESCRIPTION
**Pull request by AleksandarSavchev**
**What this PR does / why we need it**:
This PR adds the garden provider. It also adds a skelethon for the `Hardened Gardener Shoot Guide` ruleset.

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/diki/issues/304

**Special notes for your reviewer**:
A temporary `v0.0.0` version for Security Hardened Shoot Cluster ruleset has been implemented with the `garden` provider. This ruleset version will be removed when we publish version `v0.1.0`

**Release note**:

```feature user
NONE
```

### CODE REVIEW

File: docs/providers/garden.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Garden
+
+## Provider
+
+The `Garden` provider is capable of accessing a `garden` environment and running `rulesets` against it.
```
Comment (dimityrmirchev) [inline docs/providers/garden.md:5]:
```suggestion
The `Garden` provider is capable of accessing a Garden cluster environment and running `rulesets` against it.
```

File: docs/providers/garden.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+## Rulesets
+
+The `Gardener` provider implements the following `rulesets`:
```
Comment (dimityrmirchev) [inline docs/providers/garden.md:9]:
```suggestion
The `Garden` provider implements the following `rulesets`:
```

File: docs/providers/garden.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+- Hardened Gardener Shoot Guide
```
Comment (dimityrmirchev) [inline docs/providers/garden.md:10]:
```suggestion
- Security Hardened Shoot Cluster
```

File: example/config/garden.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+providers:       # contains information about known providers
+- id: garden     # unique provider identifier
+  name: "Garden" # user friendly name of the provider
+  metadata:
+    foo: bar
+  args:
+    # additionalOpsPodLabels: # pod labels that will be added to diki ops pods
+    #   foo: bar
```
Comment (dimityrmirchev) [inline example/config/garden.yaml:8]: On which pods are these labels added as additional?

File: pkg/provider/builder/garden.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+package builder
+
+import (
+	"fmt"
+	"log/slog"
+
+	"github.com/gardener/diki/pkg/config"
+	"github.com/gardener/diki/pkg/provider"
+	"github.com/gardener/diki/pkg/provider/garden"
+	"github.com/gardener/diki/pkg/provider/garden/ruleset/securityhardenedshootcluster"
```
Comment (dimityrmirchev) [inline pkg/provider/builder/garden.go:14]: Let's rename this everywhere. The current directory name seems too long
```suggestion
	"github.com/gardener/diki/pkg/provider/garden/ruleset/securityhardenedshoot"
```

File: pkg/provider/garden/provider.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	"github.com/gardener/diki/pkg/rule"
+	"github.com/gardener/diki/pkg/ruleset"
+	sharedprovider "github.com/gardener/diki/pkg/shared/provider"
+)
+
+// Provider is a Garden Cluster Provider that can
+// be used to implement rules against a garden cluster.
+type Provider struct {
+	id, name               string
+	AdditionalOpsPodLabels map[string]string
```
Comment (dimityrmirchev) [inline pkg/provider/garden/provider.go:28]: Why are these needed?

**Review by dimityrmirchev — COMMENTED** [review]
Some small comments/questions

File: example/config/garden.yaml (hunk 4 above)
Comment (AleksandarSavchev) [reply example/config/garden.yaml:8]: Hmm, I was thinking for the case where we add a ruleset that would need to create pods. But lets remove it and add it only when such a case arises.

File: pkg/provider/garden/provider.go (hunk 6 above)
Comment (AleksandarSavchev) [reply pkg/provider/garden/provider.go:28]: Same as above. I will remove it.

File: pkg/provider/garden/ruleset/securityhardenedshoot/ruleset.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+var _ ruleset.Ruleset = &Ruleset{}
+
+// Ruleset implements Security Hardened Shoot Cluster.
+type Ruleset struct {
+	version    string
+	rules      map[string]rule.Rule
+	Config     *rest.Config
+	numWorkers int
+	args       Args
+	instanceID string
```
Comment (dimityrmirchev) [inline pkg/provider/garden/ruleset/securityhardenedshoot/ruleset.go:36]: I think that we can also remove the instanceID
```suggestion
```

**Review by dimityrmirchev — CHANGES_REQUESTED** [review]
Last change request

**Review by dimityrmirchev — APPROVED** [review]
/lgtm
